### ISSUE TITLE
Some questions about training

### ISSUE THREAD

**Issue by Ruziy** [microsoft/mattergen#82]
Hello, developers, I have taken it upon myself to prepare a model that is as close as possible to the physical characteristics that are present in mp_20, but not absent in alex_mp_20, such as the Poisson's ratio. But at the first stage, I ran into two problems.

The problem is related to the relationship between mattergen and Linux, since my main computer runs on Windows, I had to deploy a virtual machine on Linux, but most of the virtual machines do not support gpu functionality well or do not support it at all. Based on this, the question arose whether there is an alternative installation of the necessary packages for Windows, mainly about mattergen.
After analyzing the mp_20 dataset, I came across a problem related to the lack of instances for the physical characteristics of dft_poisson_ratio 3409 with a non-zero float64 value
FORM 3409/27136, which will lead to a decrease in the quality of the material in the future. In turn, for another mechanical characteristic in alex_mp_20 ml_bulk_modulus 607683, the float64 value is not zero. Based on this, the second question arises: is there an alternative dataset for this kind of task or some idea for their solution? Thanks
upd. alex_mp_20 uses the dft_band_gap parameter, which has metrics 38950/607683, which is also significantly low in relation to the overall shape, how did you solve this problem (after all, you use the dft_band_gap model to generate)?

**Comment by danielzuegner** [comment]
Hi @Ruziy,

We discuss how we train models on partially labeled datasets in Section 2.4 of the paper ("Property-guided design"). In short: we finetune the model only on the subset of structures for which the property is actually present. Could you read the section and report back in case you still have questions?

**Comment by Ruziy** [comment]
Hi @danielzuegner , thank you for your reply, I will definitely read this. During the learning process, I had a couple more questions. The first one concerns Multi-property fine-tuning, let me remind you that I am training a model on the mp_20 dataset and based on the training results, I would like to see a model with the ability to generate a model using 2 parameters: chemical element + dft_poisson_ratio. As far as I understand, do I need to perform Multi-property fine-tuning for this? The second question is about the learning process itself. I started the training according to mattergen-train data_module=mp_20~trainer.logger. And by the 64th epoch, I have the following metrics:Epoch 64:  35%|▎| 600/1696 [14:27<26:24,  1.45s/it, v_num=0, loss_train_step=0.516, pos_train_step=3.660, cell_train_step=0.0379, atomic_numbers_train_step=0.112, loss_train_epoch=0.480, pos_train_epoch=3.090, cell_train_epoch=0.0453. Could you tell me or send a reference to your article, which describes the thresholds for good generation of material relative to metrics, or analyze my achievements, since loss_train_epoch=0.480 can in principle be considered quite tolerable, which cannot be said about pos_train_step, I would like to learn more about the metrics themselves, and what they are based on. There is also a question in your ReadMe about 360 epochs that will lead to val_loss = 0.4. But in my logs, in principle, there is no mention of val_loss / only train_loss. Thanks.

**Comment by Ruziy** [comment]
Complementing the question about training, I interrupted the training of the mp_20 model at epoch 100, achieving 0.45 loss_val (according to the checkpoint name). At the output, I got a model without input parameters to generate, let me remind you that my ultimate goal is to get a model with parameters: chemical element + dft_poisson_ratio, as described above. If my assumptions are correct and fine-tuning the parameters will make it possible to select them during generation (if not, please correct me), then elements + dft_poisson_ratio are custom features and they need to implement their own yaml, for dft_poisson_ratio this is float data, there should be no problems for them, but as for mp_20["element"], then the data has a complex appearance, like this ['C', 'Li', 'N', 'S'], which indirectly intersects with alex_mp_20["chemical_system"] but in a different format, is it possible to offer your own version for the yaml script for elements?

**Comment by danielzuegner** [comment]
Hi @Ruziy,

Let me try to take your questions one by one. 

> As far as I understand, do I need to perform Multi-property fine-tuning for this?    

Yes.

> Could you tell me or send a reference to your article, which describes the thresholds for good generation of material relative to metrics, or analyze my achievements, since loss_train_epoch=0.480 can in principle be considered quite tolerable, which cannot be said about pos_train_step, I would like to learn more about the metrics themselves, and what they are based on.

It is not possible to conclude from the training or validation loss whether the model is good or not. The only way to know is sampling and evaluating the generated materials.

> for mp_20["element"], the data is complex, like this ['C', 'Li', 'N', 'S'], which indirectly intersects with alex_mp_20["chemical system"], but in a different format

If it's identical to `chemical_system` but formatted differently, my recommendation would be to reformat the data to fit the `chemical_system` format. This way, you can avoid having to write code for handling the conditions yourself.

**Comment by Ruziy** [comment]
Hi @danielzuegner, the problem is that my model has already been trained on the ["el", "el_2"] format data from mp_20, so do I need to perform a complete retraining of the model? By first changing the data format(reform+cvs_to_dataset) to alex_mp_20["chemical system"], it's just very time-consuming.

**Comment by danielzuegner** [comment]
Hi @Ruziy,

If your model was already successfully trained with the element information, then this means that you've successfully integrated this new property to the model, and then there is nothing left to do. 

The problem with validation loss is that adding properties will make the losses incomparable to one another. For instance, adding the chemical system as a condition makes the prediction of the atom types much, much easier for the model, leading to greatly reduced atomic number loss. This means we cannot compare the loss of this model to an unconditional model's loss.

**Comment by Ruziy** [comment]
Hi @danielzuegner , I probably didn't describe my problem very well, I'll try to tell you in more detail and you could answer more precisely. My main task is to train the model on the mp_20 dataset and fine-tune it with custom parameters mp_20["elements"] and mp_20["dft_poisson_ratio"]. I have already trained the data on the original mp_20 dataset without changes mp_20["elements"]. But I ran into the problem of having to write 2 yaml scripts for fine-tuning, since mp_20["dft_poisson_ratio"] is a float type, there are no problems with it, but mp_20["elements"] has a complex architecture of the format ['C', 'Li', 'N', 'S']. From the options for solving the problem of writing yaml for mp_20["elements"]: 1. Option is to write your own code for processing the conditions (which will probably cause more problems than solutions). 2. Option is to retrain the model. If you choose the second option for the development of the situation, then you need to perform the following steps to bring mp_20["elements"] from the format ['C', 'Li', 'N', 'S'] to the format C-Li-N-S. Here the question is about the need to use csv-to-dataset because the basic architecture of mp_20 will be changed. Then train the model on the new dataset and perform fine-tuning according to alex_mp_20["chemical system"] for mp_20["elements"], and alex_mp_20["dft_mag_density"] (as an example for float) for mp_20["dft_poisson_ratio"]. I hope I was able to convey my thoughts and the vector of development clearly enough. Thank you

**Comment by danielzuegner** [comment]
I'm sorry, I don't understand your issue.

> I have already trained the data on the original mp_20 dataset without changes mp_20["elements"]

What model did you train? An unconditional model on the original unchanged mp_20 dataset? If yes, there's no need to retrain because you can simply add your new properties (chemical system & poisson ratio) during fine-tuning.

**Comment by Ruziy** [comment]
@danielzuegner, I apologize, I probably explain it terribly. Yes, everything is correct, I trained an unconditional model on the mp_20 dataset without changes. At this stage, I will try to be more precise, I suggest forgetting about the Poisson coefficient because there should be no problems with it. Let's stop at mp_20["elements"], my goal is to fine-tune the model so that I can specify the chemical elements of the material when generating. To do this, you need to write yaml, but for the data format from mp_20["elements"] like ['C', 'Li', 'N', 'S'] there are no implemented yamls, that's my problem. As a solution, we came to the fact that I can use chemical_system.yaml . But here a question arises since there is a difference in formats between mp_20["elements"]=>list and alex_mp_20["chemical system"]=>str, this will probably lead to errors. Developing this idea, my unconditional model is also trained on the format where mp_20["elements"]=>list, then it will have to be retrained, having previously changed mp_20["elements"]=>list_to_str=>like alex_mp_20["chemical system"] for chemical_system.yaml to work, right? Correcting you, I expect the input parameters elements+dft_poisson_ratio, not chemical_system+dft_poisson_ratio, in simple words, I do not take data from alex_mp_20. I hope I was more precise now.

**Comment by Ruziy** [comment]
Developing the idea about your proposal chemical system & poisson ratio) during fine-tuning, how can I do this since chemical system is not in mp_20? And it cannot be used according to mattergen/conf/data_module/mp_20.yaml .
# Supported properties:
# - dft_bulk_modulus
# - dft_band_gap
# - dft_mag_density
